File: clojure/bob/bob.clj
(ns bob)

(defn silent? [prompt]
  (re-matches #"^\s*$" prompt))

(defn all-caps? [prompt]
  (re-matches #"^[A-Z\d[^\w]]+$" prompt))

(defn contains-letter? [prompt]
  (re-matches #".*[A-Z].*" prompt))

(defn yelling? [prompt]
  (and (contains-letter? prompt)
       (all-caps? prompt)))

(defn question? [prompt]
  (= (last prompt) \?))

(defn response-for [prompt]
  (cond
    (yelling? prompt) "Whoa, chill out!"
    (question? prompt) "Sure."
    (silent? prompt) "Fine. Be that way!"
    :else "Whatever."))

File: javascript/bob/bob.js
//
// This is only a SKELETON file for the "Bob" exercise. It's been provided as a
// convenience to get you started writing code faster.
//

var Bob = function() {};

Bob.prototype.hey = function(input) {
  if (isShouting(input)) {
    return 'Whoa, chill out!';
  }
  if (isQuestion(input)) {
    return 'Sure.';
  }
  if (isSilence(input)) {
    return 'Fine. Be that way!';
  }
  return 'Whatever.';
};

function isShouting(input) {
  return input.toUpperCase() === input && input.match(/[A-Za-z]+/);
}

function isQuestion(input) {
  return input.slice(-1) === '?';
}

function isSilence(input) {
  return input.match(/^\s*$/);
}

module.exports = Bob;
